File: HW2_EKF_SLAM/EKF_slam/EKF_SLAM.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  16833 Robot Localization and Mapping  % 
%  Assignment #2                         %
%  EKF-SLAM                              %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clear;
close all;
clc;

%==== TEST: Setup uncertainity parameters (try different values!) ===
sig_x = 0.25;
sig_y = 0.1;
sig_alpha = 0.1;
sig_beta = 0.01;
sig_r = 0.08;

%==== Generate sigma^2 from sigma ===
sig_x2 = sig_x^2;
sig_y2 = sig_y^2;
sig_alpha2 = sig_alpha^2;
sig_beta2 = sig_beta^2;
sig_r2 = sig_r^2;

%==== Open data file ====
fid = fopen('../data/data.txt');

%==== Read first measurement data ====
tline = fgets(fid);
arr = str2num(tline);
measure = arr';
t = 1;
 
%==== Setup control and measurement covariances ===
control_cov = diag([sig_x2, sig_y2, sig_alpha2]);
measure_cov = diag([sig_beta2, sig_r2]);

%==== Setup initial pose vector and pose uncertainty ====
pose = [0 ; 0 ; 0];
pose_cov = diag([0.02^2, 0.02^2, 0.1^2]);

%==== TODO: Setup initial landmark vector landmark[] and covariance matrix landmark_cov[] ====
%==== (Hint: use initial pose with uncertainty and first measurement) ====

% Write your code here...
[num_measure, ~] = size(measure);
landmark = [];
landmark_cov = [];
for k = 1 : num_measure / 2
    beta = measure(2 * k - 1);
    r = measure(2 * k);
    [dx, dy] = pol2cart(beta + pose(3), r);
    r_vec = [dx ; dy];
    landmark_k = pose(1 : 2) + r_vec;
    landmark = [landmark ; landmark_k];
    
    landmark_pose = [1, 0, -r * sin(pose(3) + beta); 0, 1, r * cos(pose(3) + beta)];
    landmark_measure = [-r * sin(pose(3) + beta), cos(pose(3) + beta) ; r * cos(pose(3) + beta), sin(pose(3) + beta)];
    landmark_cov = blkdiag(landmark_cov, landmark_pose * pose_cov * landmark_pose' + landmark_measure * measure_cov * landmark_measure');
end

%==== Setup state vector x with pose and landmark vector ====
x = [pose ; landmark];

%==== Setup covariance matrix P with pose and landmark covariances ====
P = [pose_cov zeros(3, 2*k) ; zeros(2*k, 3) landmark_cov];

%==== Plot initial state and conariance ====
last_x = x;
drawTrajAndMap(x, last_x, P, 0);

%==== Read control data ====
tline = fgets(fid);
while ischar(tline)
    arr = str2num(tline);
    d = arr(1);
    alpha = arr(2);
    
    %==== TODO: Predict Step ====
    %==== (Notice: predict state x_pre[] and covariance P_pre[] using input control data and control_cov[]) ====
    
    % Write your code here...
    
    input = [d ; 0 ; alpha];
    x_pre = x + [eul2rotm([0 0 x(3)] , 'XYZ') * input ; zeros(2 * k , 1)];
    
    [tx, ty] = pol2cart(-x(3) , input(1));
    state_mat = makehgtform('translate', [ty tx 0]);
    state_mat(3,:) = [];
    state_mat(:,3) = [];
    B = [cos(x(3)), -sin(x(3)), 0; sin(x(3)), cos(x(3)), 0; 0, 0, 1];
    input_mat = eul2rotm([0 0 x(3)] , 'XYZ');
    state_jacobi = blkdiag(state_mat, eye(2 * k));
    %P_pre = P;
    %P_pre(1 : 3, 1 : 3) = state_mat * P(1 : 3, 1 : 3) * state_mat' + input_mat * control_cov * input_mat';
    P_pre = state_jacobi * P * state_jacobi'; 
    P_pre(1 : 3, 1 : 3) = P_pre(1 : 3, 1 : 3) + input_mat * control_cov * input_mat';  
    
    %==== Draw predicted state x_pre[] and covariance P_pre[] ====
    drawTrajPre(x_pre, P_pre);
    
    %==== Read measurement data ====
    tline = fgets(fid);
    arr = str2num(tline);
    measure = arr';
    
    %==== TODO: Update Step ====
    %==== (Notice: update state x[] and covariance P[] using input measurement data and measure_cov[]) ====
    
    % Write your code here...
    measure_jacobi = [];
    measure_pre = [];
    Q = [];
    for i = 1 : k
        landmark_i = x_pre(3 + 2 * i - 1 : 3 + 2 * i);
        r_vec = landmark_i - x_pre(1 : 2);
        
        measure_jacobi = [measure_jacobi ; zeros(2, 3 + 2 * k)];
        measure_jacobi(end - 1, 1 : 2) = fliplr(r_vec')/norm(r_vec)^2;
        measure_jacobi(end - 1, 2) = -measure_jacobi(end - 1, 2);
        measure_jacobi(end, 1 : 2) = -r_vec'/norm(r_vec);
        measure_jacobi(end - 1 : end, 3) = [-1 ; 0];
        measure_jacobi(end - 1 : end, 3 + 2 * i - 1 : 3 + 2 * i) = -measure_jacobi(end - 1 : end, 1 : 2);
        
        [ang, r] = cart2pol(r_vec(1),r_vec(2));
        measure_pre = [measure_pre ; wrapToPi(ang - x_pre(3)) ; r];
        Q = blkdiag(Q, measure_cov);
    end
    kalman_gain = P_pre * measure_jacobi' * inv(measure_jacobi * P_pre * measure_jacobi' + Q);

    x = x_pre + kalman_gain * (measure - measure_pre);
    P = (eye(size(P)) - kalman_gain * measure_jacobi) * P_pre;
    
    %==== Plot ====   
    drawTrajAndMap(x, last_x, P, t);
    last_x = x;
    
    %==== Iteration & read next control data ===
    t = t + 1;
    tline = fgets(fid);
end

%==== EVAL: Plot ground truth landmarks ====

% Write your code here...
ground_truth = [3 6 3 12 7 8 7 14 12 6 11 12];
euclidean = [];
mahalanobis = [];
for i = 1 : k
    plot(ground_truth(2 * i - 1), ground_truth(2 * i), 'pk', 'MarkerSize',10);
    hold on
    error = [ground_truth(2 * i - 1) - x(3 + 2 * i - 1) ; ground_truth(2 * i) - x(3 + 2 * i)];
    euclidean = [euclidean norm(error)];
    mahalanobis = [mahalanobis sqrt(error' * inv(P(3 + 2 * i - 1 : 3 + 2 * i , 3 + 2 * i - 1 : 3 + 2 * i)) * error)];
end
hold off
%==== Close data file ====
fclose(fid);

File: HW2_EKF_SLAM/EKF_slam/EKF_SLAM.py
#========================================================
# 
#   16833 Robot Localization and Mapping  
#   Assignment #2 
#   EKF-SLAM       
# 
#========================================================

import numpy as np
import scipy.linalg
import scipy.spatial
import matplotlib.pyplot as plt

from drawing_utils import draw_traj_pred, draw_traj_and_map
from scipy.spatial.distance import mahalanobis

# TEST: Setup uncertainity parameters (try different values!)
stddev_x = 0.25
stddev_y = 0.1
stddev_alpha = 0.1
stddev_beta = 0.01
stddev_r = 0.08

# Open data file and read measurement data
with open('../data/data.txt') as file:
    lines = file.readlines()

line = lines.pop(0)
measure = [float(item) for item in line.split()]
t = 0

# Setup control and measurement covariances
control_cov = np.diag([stddev_x ** 2, stddev_y ** 2, stddev_alpha ** 2])
measure_cov = np.diag([stddev_beta ** 2, stddev_r ** 2])

# Setup initial pose vector and pose uncertainty
pose = np.zeros(3)
pose_cov = np.diag([0.02 ** 2, 0.02 ** 2, 0.1 ** 2])

# Setup initial landmark vector landmark[] and covariance matrix landmark_cov[]
landmark = np.empty_like(measure)
landmark_cov = np.empty((0, 0))

for k in range(len(measure) // 2):
    beta = measure[2 * k]
    r = measure[2 * k + 1]

    dx = r * np.cos(beta + pose[2])
    dy = r * np.sin(beta + pose[2])    
    
    landmark[2 * k : 2 * k + 2] = pose[0:2] + np.array([dx, dy])
    
    landmark_pose = np.array([[1, 0, -r * np.sin(pose[2] + beta)], [0, 1, r * np.cos(pose[2] + beta)]])
    landmark_measure = np.array([[-r * np.sin(pose[2] + beta), np.cos(pose[2] + beta)], [r * np.cos(pose[2] + beta), np.sin(pose[2] + beta)]])
    landmark_cov = scipy.linalg.block_diag(landmark_cov, np.matmul(np.matmul(landmark_pose, pose_cov), landmark_pose.T) \
                                           +np.matmul(np.matmul(landmark_measure, measure_cov), landmark_measure.T))

# Setup state vector x with pose and landmark vector
state_vec = np.hstack((pose, landmark))

# Setup estimate covariance matrix with pose and landmark covariances
estimate_cov = scipy.linalg.block_diag(pose_cov, landmark_cov)
measure_cov = np.kron(np.identity(len(measure) // 2), measure_cov)

# Plot initial state and covariance
state_vec_prev = np.empty_like(state_vec)
np.copyto(state_vec_prev, state_vec)
draw_traj_and_map(state_vec, state_vec_prev, estimate_cov, t)

# Read control data
while len(lines) > 0:
    # Iteration & read next control data
    t += 1
    line = lines.pop(0)
    d, alpha = [float(item) for item in line.split()]
    
    # Predict Step    
    control_vec = np.array([d, 0, alpha])
    control_jac = scipy.spatial.transform.Rotation.from_euler('z', state_vec[2], degrees=False).as_matrix()    

    state_vec_pred = np.array(state_vec)
    state_vec_pred[0:3] = state_vec_pred[0:3] + np.matmul(control_jac, control_vec)
    
    tx = d * np.cos(-state_vec[2])
    ty = d * np.sin(-state_vec[2])

    state_jac = np.array([[1, 0, ty], [0, 1, tx], [0, 0, 1]])
    state_jac = scipy.linalg.block_diag(state_jac, np.identity(len(measure)))
    
    estimate_cov_pred = np.matmul(np.matmul(state_jac, estimate_cov), state_jac.T)
    estimate_cov_pred[0:3, 0:3] = estimate_cov_pred[0:3, 0:3] + np.matmul(np.matmul(control_jac, control_cov), control_jac.T)
    
    # Draw predicted state and estimate covariance
    draw_traj_pred(state_vec_pred, estimate_cov_pred)
    
    # Read measurement data
    line = lines.pop(0)
    measure = [float(item) for item in line.split()]
    
    # Update Step
    measure_jac = np.empty((0, 3 + len(measure)))
    measure_pred = []
    
    for i in range(len(measure) // 2):
        
        dx, dy = state_vec_pred[3 + 2 * i : 5 + 2 * i] - state_vec_pred[0:2]

        r = np.hypot(dx, dy)
        beta = np.arctan2(dy, dx)
        
        measure_jac = np.vstack((measure_jac, np.zeros((2, 3 + len(measure)))))
        measure_jac[-2, 0:2] = [dy / r / r, -dx / r / r]
        measure_jac[-1, 0:2] = [-dx / r, -dy / r]
        measure_jac[-2:, 2] = [-1, 0]
        measure_jac[-2:, 3 + 2 * i : 5 + 2 * i] = -measure_jac[-2:, 0:2]

        measure_pred.append((beta - state_vec_pred[2] + np.pi) % (2 * np.pi) - np.pi)
        measure_pred.append(r)

    measure = np.array(measure)
    measure_pred = np.array(measure_pred)
    
    kalman_gain = estimate_cov_pred @ (measure_jac.T) @ scipy.linalg.inv(measure_jac @ estimate_cov_pred @ (measure_jac.T) + measure_cov)

    state_vec = state_vec_pred + kalman_gain @ (measure - measure_pred)
    estimate_cov = (np.identity(estimate_cov.shape[0]) - kalman_gain @ measure_jac) @ estimate_cov_pred
    
    # Plot
    
    draw_traj_and_map(state_vec_prev, state_vec, estimate_cov, t)
    state_vec_prev[:] = state_vec

plt.xlim(-10, 25)
plt.ylim(-5, 30)
plt.axis('equal')

# EVAL: Plot ground truth landmarks

ground_truth = np.array([[3, 6], [3, 12], [7, 8], [7, 14], [12, 6], [11, 12]])

for i in range(ground_truth.shape[0]):
    plt.scatter(ground_truth[i, 0], ground_truth[i, 1], c='black', marker=(5, 1), linewidths=2)
    print('Error for landmark' + str(i + 1))
    distance = scipy.spatial.distance.mahalanobis(ground_truth[i], state_vec[3 + 2 * i : 5 + 2 * i], scipy.linalg.inv(estimate_cov[3 + 2 * i : 5 + 2 * i, 3 + 2 * i : 5 + 2 * i]))
    print('Mahalanobis: ' + str(distance))
    distance = scipy.spatial.distance.euclidean(ground_truth[i], state_vec[3 + 2 * i : 5 + 2 * i])
    print('Euclidean: ' + str(distance))

plt.show()
